Add support for setting the group on a unix domain socket

ayush933 opened the pull request:
Add new optional, immutable string config called `unixsocketgroup`. 
Change the group of the unix socket to `unixsocketgroup` after `bind()` if specified.

Manually verified that group is modified as expected and fails on illegal arguments (like non existent group name).

Fixes #873.

ayush933 commented:
Hi, @madolson can you please take a look?

madolson commented on src/server.c:
```diff
@@ -2787,6 +2787,9 @@ void initListeners(void) {
         listener->bindaddr_count = 1;
         listener->ct = connectionByType(CONN_TYPE_UNIX);
         listener->priv = &server.unixsocketperm; /* Unix socket specified */
+        if(server.unixsocketgroup != NULL){
+            listener->group = server.unixsocketgroup; /* Unix socket group specified */
+        }
```
```suggestion
        listener->group = server.unixsocketgroup;
```
If it's not set it's NULL, so it should be okay to always just set this. I prefer this so it's clearer what the value is.

madolson commented on src/anet.c:
```diff
@@ -514,6 +515,21 @@ static int anetListen(char *err, int s, struct sockaddr *sa, socklen_t len, int
+            return ANET_ERR;
+        }
+
+        if (chown(((struct sockaddr_un *)sa)->sun_path, /*owner remains same*/ -1, grp->gr_gid)) {
```
```suggestion
        if (chown(((struct sockaddr_un *)sa)->sun_path, /*owner remains same*/ -1, grp->gr_gid) == -1) {
```
Is the documented return value.

enjoy-binbin replied:
i don't like putting the comment inline into the code, lets start it in a newline.

madolson reviewed (commented):
It would also be nice to write a test to validate this, probably in [other](https://github.com/valkey-io/valkey/blob/unstable/tests/unit/other.tcl#L410) or misc. We can validate both the permissions and group.

enjoy-binbin commented on src/connection.h:
```diff
@@ -145,6 +145,7 @@ struct connListener {
     int port;
     ConnectionType *ct;
     void *priv; /* used by connection type specified data */
+    char *group; /* used to specify unix socket group */
```
i see we put unixsocketperm into priv, and here we are adding a new group field? i suppose it is ok since the code is easy to write/read.

madolson replied:
I was also unsure about this. The other idea I had was to make it priv1/priv2 to just keep it simple. I don't really want to allocate extra memory just to put in a private field.

enjoy-binbin replied:
> I don't really want to allocate extra memory just to put in a private field.

yes, i agree. let's use priv1/priv2

ayush933 commented:
Addressed reviews and added UT to test that permissions are set correctly.
Not sure which group to use in tests since using the default root group or creating new group would require sudo.

madolson commented on tests/unit/other.tcl:
```diff
@@ -544,3 +544,12 @@ start_server {tags {"other external:skip"}} {
+        set permissions [string range [dict get $attributes -permissions] end-2 end]
+        assert_equal 744 $permissions
+    }
+}
```
```suggestion
set group [lindex [exec groups [exec whoami]] 0]
if {group != ""} {
    start_server [list tags {"repl external:skip"} overrides [list {unixsocketperm} {744} {unixsocketgroup} $group]] {
        test {test unixsocket options are set correctly} {
            set socketpath [lindex  [r config get unixsocket] 1]
            set attributes [file attributes $socketpath]
            set permissions [string range [dict get $attributes -permissions] end-2 end]
            assert_equal 744 $permissions
        }
    }
}
```
something like this might work.

ayush933 replied:
So the test were failing on ubuntu using this method with error
```
unixsocket /home/runner/work/valkey/valkey/tests/tmp/server.3167.13/socket
unixsocketgroup runner
.
.
8029:M 17 Aug 2024 17:25:19.958 # Failed opening Unix socket: chown: Operation not permitted
```
Updated the test to use the group from  a temp file instead. 
Can you please check and re-run the tests?

ayush933 replied:
Marking as resolved since all the tests passed.

madolson commented on src/anet.c:
```diff
@@ -514,6 +515,22 @@ static int anetListen(char *err, int s, struct sockaddr *sa, socklen_t len, int
 
+    if (sa->sa_family == AF_LOCAL && group != NULL) {
+        struct group *grp;
+        if ((grp = getgrnam(group)) == NULL) {
+            anetSetError(err, "getgrnam: %s", strerror(errno));
+            close(s);
+            return ANET_ERR;
+        }
+
+        /* Owner of the socket remains same. */
+        if (chown(((struct sockaddr_un *)sa)->sun_path, -1, grp->gr_gid) == -1) {
+            anetSetError(err, "chown: %s", strerror(errno));
+            close(s);
+            return ANET_ERR;
+        }
```
```suggestion
        if ((grp = getgrnam(group)) == NULL) {
            anetSetError(err, "getgrnam error for group '%s': %s", group, strerror(errno));
            close(s);
            return ANET_ERR;
        }

        /* Owner of the socket remains same. */
        if (chown(((struct sockaddr_un *)sa)->sun_path, -1, grp->gr_gid) == -1) {
            anetSetError(err, "chown error for group '%s': %s", group, strerror(errno));
            close(s);
            return ANET_ERR;
        }
```
Based on the failure, was also going to suggest we add a little bit more debugging info.

ayush933 replied:
Done.

madolson commented:
@valkey-io/core-team Please vote on this small feature, we can optionally include it in Valkey 8.0 if it's ready for rc2, but not after.

enjoy-binbin commented on src/connection.h:
```diff
@@ -144,7 +144,8 @@ struct connListener {
     ConnectionType *ct;
-    void *priv; /* used by connection type specified data */
+    void *priv1; /* used by connection type specified data */
+    void *priv2;
```
```suggestion
    void *priv2; /* used by connection type specified data */
```

zuiderkwast replied:
*Two* custom type-specific fields? It looks like a hack. It makes the API a bit messy. Can't the unix socket listener embed both in the same priv field, for example in another allocated struct with two fields (perm and group)?

I don't feel strongly about it though.

enjoy-binbin replied:
see https://github.com/valkey-io/valkey/pull/901#pullrequestreview-2237090821

zuiderkwast replied:
OK!

enjoy-binbin commented on tests/unit/other.tcl:
```diff
@@ -544,3 +544,26 @@ start_server {tags {"other external:skip"}} {
+if {$group != ""} {
+    start_server [list tags {"repl external:skip"} overrides [list unixsocketgroup $group unixsocketperm 744]] {
+        test {test unixsocket options are set correctly} {
+            set socketpath [lindex  [r config get unixsocket] 1]
```
```suggestion
            set socketpath [lindex [r config get unixsocket] 1]
```

hwware commented on src/config.c:
```diff
@@ -3110,6 +3110,7 @@ standardConfig static_configs[] = {
     /* String Configs */
     createStringConfig("aclfile", NULL, IMMUTABLE_CONFIG, ALLOW_EMPTY_STRING, server.acl_filename, "", NULL, NULL),
     createStringConfig("unixsocket", NULL, IMMUTABLE_CONFIG, EMPTY_STRING_IS_NULL, server.unixsocket, NULL, NULL, NULL),
+    createStringConfig("unixsocketgroup", NULL, IMMUTABLE_CONFIG, EMPTY_STRING_IS_NULL, server.unixsocketgroup, NULL, NULL, NULL),
```
If the default group is wheel, I prefer setting it here.

ayush933 replied:
wheel is a privileged group on most systems and not the default.

hwware commented on src/connection.h:
```diff
@@ -144,7 +144,8 @@ struct connListener {
     ConnectionType *ct;
-    void *priv; /* used by connection type specified data */
+    void *priv1; /* used by connection type specified data */
+    void *priv2; /* used by connection type specified data */
```
IMO, I do not like name the variable name as priv1, priv2.  Later, we will not know the meaning they represent.  I believe priv1 represents unixsocketperm, and thus maybe void *group or ugroup is better than now to represent unixsocketgroup

ayush933 replied:
see https://github.com/valkey-io/valkey/pull/901#pullrequestreview-2237090821

madolson reviewed (approved):
LGTM, just waiting on majority approval

madolson commented:
Since the new test is a bit odd, running the daily against a bunch of different platforms. https://github.com/valkey-io/valkey/actions/runs/10530596918
